### ISSUE TITLE
Incorrect query result for `MATCH`/`OPTIONAL MATCH` combination

### ISSUE THREAD

**Issue by sapalli2989** [kuzudb/kuzu#2942]
```cypher
CREATE NODE TABLE V (id STRING, PRIMARY KEY(id));
CREATE NODE TABLE R (id STRING, PRIMARY KEY(id));
CREATE REL TABLE childof (FROM V TO V);
CREATE REL TABLE relatedto (FROM R TO V);

CREATE (v1:V {id:"v1"}),(v11:V {id:"v11"}),(v12:V {id:"v12"}),(v111:V {id:"v111"}),
(r1:R {id:"r1"}), (r12:R {id:"r12"}),
(v111)-[:childof]->(v11),(v11)-[:childof]->(v1),(v12)-[:childof]->(v1),
(r1)-[:relatedto]->(v1), (r12)-[:relatedto]->(v12);
```
```cypher
MATCH (v1:V {id: "v1"})
WITH v1
MATCH (v:V)-[e:childof*]->(v1)
OPTIONAL MATCH (r)-[:relatedto]->(v)
RETURN r, v;
```
![screenshot](https://github.com/kuzudb/kuzu/assets/147163559/051efbf4-60b4-4efe-b227-2783fe4e8ad6)

What Kuzu returns:
```
---------------------------------------
| r | v                               |
---------------------------------------
|   | {_ID: 0:3, _LABEL: V, id: v111} |
---------------------------------------
|   | {_ID: 0:2, _LABEL: V, id: v12}  |
---------------------------------------
|   | {_ID: 0:1, _LABEL: V, id: v11}  |
---------------------------------------
(3 tuples)
(2 columns)
```

What Neo4j returns (and what is expected here):
```
╒════════════════╤═════════════════╕
│r               │v                │
╞════════════════╪═════════════════╡
│null            │(:V {id: "v11"}) │
├────────────────┼─────────────────┤
│null            │(:V {id: "v111"})│
├────────────────┼─────────────────┤
│(:R {id: "r12"})│(:V {id: "v12"}) │
└────────────────┴─────────────────┘
```
Why does Kuzu omit the existing optional match `(:R {id: "r12"})` for `v12` here?

**Comment by prrao87** [comment]
Hi @sapalli2989, we'll take a look at this as soon as possible.

**Comment by andyfengHKU** [comment]
Hi @sapalli2989 , this issue should be solved in #2977 now.
